consensus/istanbul/celo-ibft-spec.md

```diff
@@ -0,0 +1,486 @@
+// change message.
+// 
+// Note: This function is referred to in the code as
+// `handleTimeoutAndMoveToNextRound`, which is misleading because it does not move
+// to the next round, it only updates the desired round and sends a round change
+// message. Hence why it has been renamed here to avoid confusion.
+onRoundChangeTimeout(H, R) {
+  if H = Hc && R = Rc {
+    Rd ← Rc+1
+    Sc ← WaitingForNewRound
+    schedule onRoundChangeTimeout(Hc, Rd) after roundChangeTimeout(Rd)
+    bc<RoundChange, Hc, Rd, PCc>
+  }
+}
+```
+
+### Supporting Functions
+
+These functions can read global state but cannot modify it.
+
+#### Application provided functions
+No pseudocode is provided for these functions since their implementation is
+application specific.
+
+`proposer(H,R)`\
+Returns the proposer for the given height and round.
+
+`isProposer(H, R)`\
+Returns true if the current participant is the proposer for the given height
+and round. 
+
+`deliverValue(V)`\
+Delivers the given value to the application.
+
+`roundChangeTimeout(R)`\
+Returns the timeout for the given round 
+
+`bc(<PP, H, R, V>)`\
+Broadcasts the given message to all connected participants. 
+
+`send(<Commit, H, R, V>, sender(m))`\
+Sends the given message to to the sender of another message.
+
+#### PCRound
+Asserts that all messages in the given prepared certificate share the same round and returns that round.
+```
+PCRound(<PreparedCertificate, M, *>) {
+  ∃ R : ∀ m<*, *, Rm, *> ∈ M : R = Rm
+  return R
+}
+```
+
+#### PCValue
+Return the value associated with a prepared certificate.
+```
+PCValue(<PreparedCertificate, *, V>) {
+  return V
+}
+```
+
+#### validPC
+
+Returns true if the message set contains prepare or commit messages from at
+least 2f+1 and no more than 3f+1 participants for current height, matching the
+prepared certificate value and all sharing the same height and round.
+```
+validPC(<PreparedCertificate, M, V>) {
+  N ← { m<T, Hc, *, Vm> ∈ M : (T = Prepare || T = Commit) && Vm = V } 
+  return 2f+1 <= |N| <= 3f+1 &&
+  ∀ m<*, Hm, Rm, *>, n<*, Hn, Rn, *> ∈ N : Hm = Hn && Rm = Rn 
+}
+```
+
+#### validRCC
+
+Returns true if the round change contains at least 2f+1 and no more than 3f+1
+round changes for the current height, with a round greater or equal to the
+desired round and either have a valid prepared certificate or no prepared
+certificate. If any round change certificates have a prepared certificate, then
+there must exist one with round greater than or equal to all the others and
+with a value of V.
+
+```
+validRCC(V, RCC) {
+  M ← { m<RoundChange, Hm , Rm, PC> ∈ RCC : Hm = Hc && Rm >= Rd && (PC = nil || validPC(PC)) }
+  N ← { m<RoundChange, * , *, PC> ∈ M : PC != nil }
+  if |N| > 0 {
+    return 2f+1 <= |M| <= 3f+1 &&
+    ∃ m<RoundChange, *, *, Pcm> ∈ N : validPC(PCm) && PCValue(PCm) = V && ∀ n<RoundChange, * , *, PCn> ∈ N != m : PCRound(PCm) >= PCRound(PCn)
+  }
+  return 2f+1 <= |M| <= 3f+1 &&
```

Comment:
Dangling `&&`. I don't think there's a condition there that needs to be checked.

```suggestion
  return 2f+1 <= |M| <= 3f+1
```